### ISSUE TITLE
Improve the Inductor generated kernel for the pattern of `output1 = pointwise(intput); output2 = transpose(output1)`

### ISSUE THREAD

**Issue by y-sq** [pytorch/pytorch#130015]
### 🚀 The feature, motivation and pitch

**Context**:
While using [float8 training](https://fburl.com/4mblb81i), the operators of `fp8 = cast_to_fp8(input_tensor); fp8_t = fp8.t().contiguous().t()` is used in every float8 linear. We use Inductor to generate the performant fused kernel for these ops. However, there are some inefficiencies in the generated kernels.

**Examples**
Following is an example of the pattern. And two issues are noticed from the two test functions.
```
def test() -> None:
    ref_dtype = torch.bfloat16
    M, K, N = 4096, 4096, 3072
    
    input_tensor = torch.randn(M, K, device="cuda", dtype=ref_dtype, requires_grad=False)
    scale = torch.Tensor([10.0]).to("cuda")

    E4M3_MAX_POS = torch.finfo(torch.float8_e4m3fn).max
    E5M2_MAX_POS = torch.finfo(torch.float8_e5m2).max

    def test_pattern1(tensor_x_inp, scale_x):
        tensor_x = tensor_x_inp * scale_x
        tensor_x = tensor_x.clamp(min=-1 * E4M3_MAX_POS, max=E4M3_MAX_POS)
        tensor_fp8 = tensor_x.to(torch.float8_e4m3fn)

        tensor_fp8_t = tensor_fp8.t().contiguous().t()

        return (tensor_fp8, tensor_fp8_t)

    test_pattern = torch.compile(test_pattern1)
    tensor_fp8, tensor_fp8_t = test_pattern(input_tensor, scale)
    print(tensor_fp8.stride(), tensor_fp8_t.stride())

    def test_pattern2(tensor_x_inp, scale_x):
        tensor_x = tensor_x_inp * scale_x
        tensor_x = tensor_x.clamp(min=-1 * E4M3_MAX_POS, max=E4M3_MAX_POS)
        tensor_fp8 = tensor_x.to(torch.float8_e4m3fn)

        tensor_x_t = (tensor_x_inp * scale_x).t()
        tensor_x_t = tensor_x_t.clamp(min=-1 * E4M3_MAX_POS, max=E4M3_MAX_POS)
        tensor_fp8_t = tensor_x_t.to(torch.float8_e4m3fn)
        
        tensor_fp8_t = tensor_fp8_t.contiguous().t()

        return (tensor_fp8, tensor_fp8_t)

    test_pattern = torch.compile(test_pattern2)
    tensor_fp8, tensor_fp8_t = test_pattern(input_tensor, scale)
    print(tensor_fp8.stride(), tensor_fp8_t.stride())
```

**The issue in "test_pattern1"**
In `test_pattern1`, ` tensor_fp8.t().contiguous().t()` is not fused with the casting kernel. Instead, Inductor generated two kernels (`triton_poi_fused__to_copy_clamp_mul_0`, and `triton_poi_fused_clone_1`). 

From Inductor log, 
```
V0702 16:39:45.362000 139774580315136 torch/_inductor/scheduler.py:1842] [0/0] [__fusion] ===== attempting fusion (1/10): 2 nodes =====
V0702 16:39:45.362000 139774580315136 torch/_inductor/scheduler.py:2084] [0/0] [__fusion] fuse_nodes_once, candidates:
V0702 16:39:45.362000 139774580315136 torch/_inductor/scheduler.py:2086] [0/0] [__fusion]   SchedulerNode(name='buf0'), Pointwise(['[4096, 4096]', 'origins={mul, convert_element_type, clamp_max, clamp_min}'])
V0702 16:39:45.362000 139774580315136 torch/_inductor/scheduler.py:2086] [0/0] [__fusion]   SchedulerNode(name='buf1'), Pointwise(['[4096, 4096]', 'origins={clone}'])
V0702 16:39:45.362000 139774580315136 torch/_inductor/scheduler.py:645] [0/0] [__fusion] cannot fuse buf0 with buf1: no shared data
```

**The issue in "test_pattern2"**
If we use `test_pattern2`, only one fused kernel (`triton_poi_fused__to_copy_clamp_clone_mul_0`) is generated. However, the bandwidth of the fused kernel is only 1.5 TB/s (less than the expected 2 TB/s on an H100). 
Looking at the generated kernel,
```
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code] @triton_heuristics.pointwise(
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     size_hints=[4096, 4096], tile_hint=TileHint.DEFAULT,
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     filename=__file__,
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     triton_meta={'signature': {0: '*bf16', 1: '*fp32', 2: '*fp8e4nv', 3: '*fp8e4nv', 4: 'i32', 5: 'i32'}, 'device': DeviceProperties(type='cuda', index=0, cc=90, major=9, regs_per_multiprocessor=65536, max_threads_per_multi_processor=2048, multi_processor_count=132), 'constants': {}, 'configs': [AttrsDescriptor(divisible_by_16=(0, 1, 2, 3, 4, 5), equal_to_1=())]},
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     inductor_meta={'autotune_hints': set(), 'kernel_name': 'triton_poi_fused__to_copy_clamp_clone_mul_0', 'mutated_arg_names': [], 'no_x_dim': False, 'num_load': 3, 'num_reduction': 0, 'backend_hash': '111C3DDC11B3C29F74BF1795749D65D0F93E66347392E755AD82B8A40774FD60', 'are_deterministic_algorithms_enabled': False, 'assert_indirect_indexing': True, 'autotune_local_cache': True, 'autotune_pointwise': True, 'autotune_remote_cache': False, 'force_disable_caches': True, 'dynamic_scale_rblock': True, 'max_autotune': False, 'max_autotune_pointwise': False, 'min_split_scan_rblock': 256, 'spill_threshold': 16, 'store_cubin': False, 'profile_bandwidth': True, 'profile_bandwidth_regex': '', 'profile_bandwidth_output': '/tmp/profile.txt', 'kernel_num_gb': 0.067108868},
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     min_elem_per_thread=2
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code] )
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code] @triton.jit
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code] def triton_(in_ptr0, in_ptr1, out_ptr0, out_ptr1, ynumel, xnumel, YBLOCK : tl.constexpr, XBLOCK : tl.constexpr):
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     ynumel = 4096
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     xnumel = 4096
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     yoffset = (tl.program_id(1) + tl.program_id(2) * tl.num_programs(1)) * YBLOCK
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     yindex = yoffset + tl.arange(0, YBLOCK)[None, :]
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     ymask = tl.full([XBLOCK, YBLOCK], True, tl.int1)
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     xoffset = tl.program_id(0) * XBLOCK
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     xindex = xoffset + tl.arange(0, XBLOCK)[:, None]
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     xmask = tl.full([XBLOCK, YBLOCK], True, tl.int1)
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     x1 = xindex
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     y0 = yindex
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     tmp0 = tl.load(in_ptr0 + (x1 + (4096*y0)), None).to(tl.float32)
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     tmp2 = tl.load(in_ptr1 + (0))
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     tmp3 = tl.broadcast_to(tmp2, [XBLOCK, YBLOCK])
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     tmp10 = tl.load(in_ptr0 + (y0 + (4096*x1)), None).to(tl.float32)
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     tmp1 = tmp0.to(tl.float32)
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     tmp4 = tmp1 * tmp3
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     tmp5 = -448.0
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     tmp6 = triton_helpers.maximum(tmp4, tmp5)
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     tmp7 = 448.0
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     tmp8 = triton_helpers.minimum(tmp6, tmp7)
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     tmp9 = tmp8.to(tl.float8e4nv)
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     tmp11 = tmp10.to(tl.float32)
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     tmp12 = tmp11 * tmp3
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     tmp13 = triton_helpers.maximum(tmp12, tmp5)
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     tmp14 = triton_helpers.minimum(tmp13, tmp7)
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     tmp15 = tmp14.to(tl.float8e4nv)
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     tl.store(out_ptr0 + (x1 + (4096*y0)), tmp9, None)
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code]     tl.store(out_ptr1 + (x1 + (4096*y0)), tmp15, None)
V0702 16:39:46.770000 139774580315136 torch/_inductor/graph.py:1680] [1/0] [__output_code] ''', device_str='cuda')
```
One simple change that can be applied to the kernel is to remove the load of `tmp10`, and write casted `tmp0` to the transposed position of `out_ptr1`.

### Alternatives

For a short-term fix, we added a handwritten kernel for the fused cast+transpose, and use Inductor pattern matcher to replace the ops with the handwritten kernel. But it would be nice to have a general fix in Inductor.

**Comment by shunting314** [comment]
> One simple change that can be applied to the kernel is removing the load of tmp10, and write casted tmp0 to the transposed position of out_ptr1.

Can you share the change and the benchmark result?

**Comment by y-sq** [comment]
> > One simple change that can be applied to the kernel is removing the load of tmp10, and write casted tmp0 to the transposed position of out_ptr1.
> 
> Can you share the change and the benchmark result?

Hi @shunting314 , I had the changed kernel and benchmark result documented in a fb-internal document which you should have access to, https://fburl.com/gdoc/gtacaqlt

**Comment by shunting314** [comment]
For the not able to fusion problem you illustrate in test_pattern1, I've verified that our WIP loop ordering after fusion PR can handle it properly and I've added this as a test here: https://github.com/pytorch/pytorch/pull/126254/files#diff-1729825ff86b624333e2fc22c4a947b8526757e703787eb0fc1b607f3c3b8705R374 

But we that PR is still WIP and we don't want to rush for this fundamental feature.

I'll think if I can come up with some short term hack before loop ordering after fusion is landed.

**Comment by shunting314** [comment]
One hack to solve the fusion problem here is to flip the order here https://github.com/pytorch/pytorch/blob/cd70ac884f25762c70678617b6acad94c9bb4d74/torch/_inductor/scheduler.py#L1370 when there is a tie.  This can be controlled by a config. But I'm not sure if it's worth to have such a hack or we should wait for LoopOrderAfterFusion to formally ready. cc @Chillee @jansel @eellison for thoughts

**Comment by shunting314** [comment]
@y-sq with loop order after fusion or with the hack I mentioned above, Inductor would have the issue you mentioned in pattern2. Check the generated kernel here:

```
@triton.jit
def triton_poi_fused__to_copy_clamp_clone_mul_0(in_ptr0, in_ptr1, out_ptr0, out_ptr1, ynumel, xnumel, YBLOCK : tl.constexpr, XBLOCK : tl.constexpr):
    ynumel = 4096
    xnumel = 4096
    yoffset = (tl.program_id(1) + tl.program_id(2) * tl.num_programs(1)) * YBLOCK
    yindex = yoffset + tl.arange(0, YBLOCK)[None, :]
    ymask = tl.full([XBLOCK, YBLOCK], True, tl.int1)
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:, None]
    xmask = tl.full([XBLOCK, YBLOCK], True, tl.int1)
    x1 = xindex
    y0 = yindex
    tmp0 = tl.load(in_ptr0 + (x1 + (4096*y0)), None).to(tl.float32)
    tmp2 = tl.load(in_ptr1 + (0))
    tmp3 = tl.broadcast_to(tmp2, [XBLOCK, YBLOCK])
    tmp1 = tmp0.to(tl.float32)
    tmp4 = tmp1 * tmp3
    tmp5 = -448.0
    tmp6 = triton_helpers.maximum(tmp4, tmp5)
    tmp7 = 448.0
    tmp8 = triton_helpers.minimum(tmp6, tmp7)
    tmp9 = tmp8.to(tl.float8e4nv)
    tl.store(out_ptr0 + (x1 + (4096*y0)), tmp9, None)
    tl.store(out_ptr1 + (y0 + (4096*x1)), tmp9, None)
```

**Comment by jansel** [comment]
+1 for loop ordering after fusion

@shunting314 is that close to being finished?

**Comment by shunting314** [comment]
> is that close to being finished?

The main functionality is there. But I still need to debug a few CI failures and optimize more for compilation time. Still need some time. But I'm actively working on this

**Comment by shunting314** [comment]
One possible way to move forward faster is, I clean up all the failures first and have a version that still regression compilation time. We land that with loop ordering after fusion disabled by default. And then @y-sq can enable loop ordering after fusion  in her code. The only downside is extra compilation time for y-sq's use cases. At the same time, we optimize for compilation time.

But this most likely need to happen after I'm back from PTO

**Comment by vkuzo** [comment]
@shunting314 , is there a short term workaround that can be landed sooner gated with a config to unblock float8 performance testing?

also, I was wondering if someone could share context on `LoopOrderAfterFusion`, to help people less familiar with inductor understand

**Comment by vkuzo** [comment]
I wanted to share some more context on the importance of "issue 1".  Looking at GPU kernel time of a fwd + bwd through a linear converted to float8 with dynamic scaling, with M, K, N = 4096, 4096, 4096 (a medium sized linear), we currently see the following breakdown of overhead:

```
   experiment                                                                          kernel       category  time_ms  pct_gpu_time bw_gpbs
3   1_float8                                                                aten::_scaled_mm         0_gemm    0.370         0.682    None
0   1_float8                                                      triton_red_fused_abs_max_0  1_f8_overhead    0.060         0.111    None
2   1_float8                                triton_poi_fused__to_copy_clamp_mul_reciprocal_2  1_f8_overhead    0.040         0.073    None
6   1_float8          triton_poi_fused__scaled_mm__to_copy_clamp_clone_mul_reciprocal_view_2  1_f8_overhead    0.030         0.056    None
1   1_float8                        triton_per_fused__to_copy_abs_clamp_max_mul_reciprocal_1  1_f8_overhead    0.004         0.006    None
5   1_float8  triton_per_fused__scaled_mm__to_copy_abs_clamp_clone_max_mul_reciprocal_view_1  1_f8_overhead    0.002         0.003    None
4   1_float8                                                        triton_poi_fused_clone_3        2_other    0.037         0.068    None
```

The additional kernels added by "issue_1" above are represented by `triton_poi_fused_clone_3`, and they take 6.8% of total benchmark time, or 0.068 / (1 - 0.682) = 21% of the float8 overhead (subtracting the gemm).  I have observed similar magnitudes of these extra clone kernels in subgraphs extracted from Meta models.  So, fixing this should be a really nice ~5%-25% float8 overhead reduction for medium problem sizes across the board, for the cases where activation checkpointing is turned off.

**Comment by vkuzo** [comment]
Moving "issue 2" from this issue here: https://github.com/pytorch/pytorch/issues/133242

**Comment by vkuzo** [comment]
Verifying that the repro provided in this issue is to be fixed by https://github.com/pytorch/pytorch/pull/126254, and we see a speedup.  Thanks for the great work!  IMO we should leave this issue open until `TORCHINDUCTOR_LOOP_ORDERING_AFTER_FUSION` is ready for production usage.

```
// script: https://gist.github.com/vkuzo/d2faa9d363a2c9a5e2885c2781414614

// before
(pytorch) [vasiliy@devgpu006.vll6 ~/local/ao (main)]$ TORCHINDUCTOR_PROFILE=1 TORCHINDUCTOR_LOOP_ORDERING_AFTER_FUSION=0 python ~/local/tmp/repro_issue_130015.py 
0.030ms         0.050 GB         1696.73GB/s     triton_poi_fused__to_copy_clamp_mul_0
0.026ms         0.034 GB         1310.72GB/s     triton_poi_fused_clone_1
SUMMARY (/tmp/torchinductor_vasiliy/ed/ced6h7kbgvcwgbxs5id7oe2uayptp2hkb3d5vc77ls2kzyom44a2.py)
0.06ms           0.08 GB         1517.92GB/s

// after
(pytorch) [vasiliy@devgpu006.vll6 ~/local/ao (main)]$ TORCHINDUCTOR_PROFILE=1 TORCHINDUCTOR_LOOP_ORDERING_AFTER_FUSION=1 python ~/local/tmp/repro_issue_130015.py 
0.040ms         0.067 GB         1671.04GB/s     triton_poi_fused__to_copy_clamp_clone_mul_0
SUMMARY (/tmp/torchinductor_vasiliy/hz/chzdkuxbd3y4zs3c4tpgnm5qeyya5cj6k6dmboanyld4opkkmd5a.py)
0.04ms           0.07 GB         1671.04GB/s

(4096, 1) (1, 4096)
```

**Comment by vkuzo** [comment]
from @eellison , `TORCHINDUCTOR_LOOP_ORDERING_AFTER_FUSION` is still off by default, will check in on when we can turn it on.

**Comment by danielvegamyhre** [comment]
@shunting314 @eellison @vkuzo FYI using TORCHINDUCTOR_LOOP_ORDERING_AFTER_FUSION did not improve perf for rowwise scales like it did for tensorwise scales, for the internal benchmarks I ran that we discussed. Is this unexpected?

**Comment by danielvegamyhre** [comment]
@shunting314 are you still planning to work on this? This will help improve fp8 rowwise training perf

**Comment by shunting314** [comment]
@danielvegamyhre  yes, working on enabling it by default now. Sorry for the delay, was just back from a long leave.
